balancer/balanceraggregator/balanceraggregator.go

```diff
@@ -0,0 +1,286 @@
+		bal := child.(balancer.Balancer)
+		bal.Close()
+	}
+}
+
+// updateState updates this component's state. It sends the aggregated state,
+// and a picker with round robin behavior with all the child states present if
+// needed.
+func (ba *BalancerAggregator) updateState() {
+	if ba.inhibitChildUpdates {
+		return
+	}
+	readyPickers := make([]balancer.Picker, 0)
+	connectingPickers := make([]balancer.Picker, 0)
+	idlePickers := make([]balancer.Picker, 0)
+	transientFailurePickers := make([]balancer.Picker, 0)
+
+	childStates := make([]ChildState, 0, ba.children.Len())
+	for _, child := range ba.children.Values() {
+		bw := child.(*balancerWrapper)
+		bw.mu.Lock()
+		childState := bw.childState
+		bw.mu.Unlock()
+		childStates = append(childStates, childState)
+		childPicker := childState.State.Picker
+		switch childState.State.ConnectivityState {
+		case connectivity.Ready:
+			readyPickers = append(readyPickers, childPicker)
+		case connectivity.Connecting:
+			connectingPickers = append(connectingPickers, childPicker)
+		case connectivity.Idle:
+			idlePickers = append(idlePickers, childPicker)
+		case connectivity.TransientFailure:
+			transientFailurePickers = append(transientFailurePickers, childPicker)
+			// connectivity.Shutdown shouldn't appear.
+		}
+	}
+
+	// Construct the round robin picker based off the aggregated state. Whatever
+	// the aggregated state, use the pickers present that are currently in that
+	// state only.
+	var aggState connectivity.State
+	var pickers []balancer.Picker
+	if len(readyPickers) >= 1 {
+		aggState = connectivity.Ready
+		pickers = readyPickers
+	} else if len(connectingPickers) >= 1 {
+		aggState = connectivity.Connecting
+		pickers = connectingPickers
+	} else if len(idlePickers) >= 1 {
+		aggState = connectivity.Idle
+		pickers = idlePickers
+	} else if len(transientFailurePickers) >= 1 {
+		aggState = connectivity.TransientFailure
+		pickers = transientFailurePickers
+	} else {
+		aggState = connectivity.TransientFailure
+		pickers = append(pickers, base.NewErrPicker(errors.New("no children to pick from")))
+	} // No children (resolver error before valid update).
+
+	p := &pickerWithChildStates{
+		pickers:     pickers,
+		childStates: childStates,
+		next:        uint32(grpcrand.Intn(len(pickers))),
+	}
+	ba.cc.UpdateState(balancer.State{
+		ConnectivityState: aggState,
+		Picker:            p,
+	})
+}
+
+// pickerWithChildStates delegates to the pickers it holds in a round robin
+// fashion. It also contains the childStates of all the BalancerAggregator's
+// children.
+type pickerWithChildStates struct {
+	pickers     []balancer.Picker
+	childStates []ChildState
+	next        uint32
+}
+
+func (p *pickerWithChildStates) Pick(info balancer.PickInfo) (balancer.PickResult, error) {
+	nextIndex := atomic.AddUint32(&p.next, 1)
+	picker := p.pickers[nextIndex%uint32(len(p.pickers))]
+	return picker.Pick(info)
+}
+
+// ChildStatesFromPicker returns the persisted child states from the picker if
```

Comment:
Nit: "persisted" seems to add no benefit to this sentence.  And "picker if picker" reads awkwardly.  How about just:

```go
// ChildStatesFromPicker returns the most recent state all children managed by the
// BalancerAggregator <name TBD> balancer that produced this picker.
```

Reply:
Deleted persisted. added the i.e. picker if the picker. I think we need to leave the rest as is though, because your comment implies it returns the most recent state unconditionally, where it is conditional on the picker type.

Reply:
Yeah, it does, but only if the provided picker is the right type (it takes a picker interface, as that is the type provided to parent in UpdateState, not the concrete type of the picker. Otherwise there is nothing to look into for child states.

Reply:
Please add "managed by this balancer" or something like that here as discussed offline.

Reply:
Switched comment to "ChildStatesFromPicker returns the child states from the picker that are managed by this balancer."